File: 2001~2500/2433. Find The Original Array of Prefix Xor/main.py
"""
Source: https://rb.gy/w0t0o
Date: 2023/8/4
Skill:
Ref:
Runtime: 652 ms, faster than 98.69%
Memory Usage: 36.58 MB, less than 25.98%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools


class Solution:
    def findArray(self, pref: List[int]) -> List[int]:
        res, sz = [pref[0]], len(pref)
        for i in range(1, sz):
            res.append(pref[i] ^ pref[i - 1])

        return res


if __name__ == "__main__":
    s = Solution()

File: 2001~2500/2433. Find The Original Array of Prefix Xor/main.cpp
/**
 * Source: ibit.ly/kzL7J
 * Date: 2023/10/31
 * Skill:
 * Runtime: 85 ms, faster than 63.55% of C++ online submissions
 * Memory Usage: 76.20 MB, less than 83.50% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


class Solution {
public:
    vector<int> findArray(vector<int>& pref) {
        vector<int> res(pref.size());
        int presum = pref[0];
        res[0] = presum;
        for (int i = 1; i < pref.size(); i++) {
            res[i] = presum ^ pref[i];
            presum ^= res[i];
        }
        return res;
    }
};


int main() {
    Solution s;
    vector<int> nums{2,3,0,0,2};
}